File: android/src/main/java/com/jimmydaddy/imagemarker/ImageMarkerManager.java
package com.jimmydaddy.imagemarker;

import static com.facebook.drawee.backends.pipeline.Fresco.getImagePipeline;
import static com.jimmydaddy.imagemarker.base.Utils.getStringSafe;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.util.Base64;
import android.util.Log;

import com.facebook.common.references.CloseableReference;
import com.facebook.datasource.DataSource;
import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.imagepipeline.core.ImagePipelineConfig;
import com.facebook.imagepipeline.datasource.BaseBitmapDataSubscriber;
import com.facebook.imagepipeline.image.CloseableImage;
import com.facebook.imagepipeline.request.ImageRequest;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.modules.systeminfo.ReactNativeVersion;
import com.jimmydaddy.imagemarker.base.MarkImageOptions;
import com.jimmydaddy.imagemarker.base.MarkTextOptions;
import com.jimmydaddy.imagemarker.base.Position;
import com.jimmydaddy.imagemarker.base.SaveFormat;
import com.jimmydaddy.imagemarker.base.TextOptions;
import com.jimmydaddy.imagemarker.base.Utils;

import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.UUID;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;


/**
 * Created by jimmydaddy on 2017/3/6.
 */


public class ImageMarkerManager extends ReactContextBaseJavaModule {
  private ReactApplicationContext context;
  private static final String IMAGE_MARKER_TAG = "[ImageMarker]";
  private static final String BASE64 = "base64";
  public static final String NAME = "ImageMarker";


  public ImageMarkerManager(ReactApplicationContext reactContext) {
    super(reactContext);
    this.context = reactContext;
  }

  @Override
  public String getName() {
    return NAME;
  }

  private Resources getResources() {
    return this.context.getResources();
  }

  private int getDrawableResourceByName(String name) {
    return this.getResources().getIdentifier(
      name,
      "drawable",
      this.context.getPackageName());
  }

//    private Drawable getDrawableByName(String name) {
//        int drawableResId = getDrawableResourceByName(name);
//        if (drawableResId != 0) {
//            return getResources().getDrawable(getDrawableResourceByName(name));
//        } else {
//            return null;
//        }
//    }

  private Boolean isFrescoImg(String uri) {
    String base64Pattern = "^data:(image|img)/(bmp|jpg|png|tif|gif|pcx|tga|exif|fpx|svg|psd|cdr|pcd|dxf|ufo|eps|ai|raw|WMF|webp);base64,(([[A-Za-z0-9+/])*\\s\\S*)*";

    return uri.startsWith("http://") || uri.startsWith("https://") || uri.startsWith("file://") || (uri.startsWith("data:") && uri.contains("base64") && (uri.contains("img") || uri.contains("image")));
  }

  private Bitmap.CompressFormat getSaveFormat(SaveFormat saveFormat) {
    return saveFormat != null && saveFormat == SaveFormat.PNG ? Bitmap.CompressFormat.PNG : Bitmap.CompressFormat.JPEG;
  }

  private void setMaxBitmapSize(int maxSize) {
    String major = getStringSafe("major", ReactNativeVersion.VERSION);
    String minor = getStringSafe("minor", ReactNativeVersion.VERSION);
    String patch = getStringSafe("patch", ReactNativeVersion.VERSION);
    if (Integer.valueOf(major) >= 0 && Integer.valueOf(minor) >= 60 && Integer.valueOf(patch) >= 0) {
      ImagePipelineConfig config = ImagePipelineConfig.newBuilder(this.context).experiment().setMaxBitmapSize(maxSize).build();
      Fresco.initialize(this.context, config);
    }
  }

  private void markImage(
    final Bitmap bg,
    final String dest,
    MarkImageOptions opts,
    final Promise promise) {
    try {
      final String uri = opts.watermarkImage.uri;
      Log.d(IMAGE_MARKER_TAG, uri);

      if (isFrescoImg(uri)) {
        ImageRequest imageRequest = ImageRequest.fromUri(uri);
        DataSource<CloseableReference<CloseableImage>> dataSource = Fresco.getImagePipeline().fetchDecodedImage(imageRequest, context);
        Executor executor = Executors.newSingleThreadExecutor();
        dataSource.subscribe(new BaseBitmapDataSubscriber() {
          @Override
          public void onNewResultImpl(Bitmap bitmap) {
            if (bitmap != null) {
              Bitmap mark = Utils.scaleBitmap(bitmap, opts.watermarkImage.scale);

              markImageByBitmap(bg, mark, dest, opts, promise);
            } else {
              promise.reject("marker error", "Can't retrieve the file from the markerpath: " + uri);
            }
          }

          @Override
          public void onFailureImpl(DataSource dataSource) {
            promise.reject("error", "Can't request the image from the uri: " + uri, dataSource.getFailureCause());
          }
        }, executor);
      } else {
        int resId = getDrawableResourceByName(uri);
        if (resId == 0) {
          Log.d(IMAGE_MARKER_TAG, "cannot find res");
          promise.reject("error", "Can't get resource by the path: " + uri);
        } else {
          Log.d(IMAGE_MARKER_TAG, "res：" + resId);

          Resources r = this.getResources();
//                    InputStream is = r.openRawResource(resId);
          Bitmap bitmap = BitmapFactory.decodeResource(r, resId);
//                    Bitmap bitmap = BitmapFactory.decodeStream(is);
          Log.d(IMAGE_MARKER_TAG, bitmap.getHeight() + "");
          Bitmap mark = Utils.scaleBitmap(bitmap, opts.watermarkImage.scale);
          Log.d(IMAGE_MARKER_TAG, mark.getHeight() + "");

          if (bitmap != null && !bitmap.isRecycled() && opts.watermarkImage.scale != 1.0f) {
            bitmap.recycle();
            System.gc();
          }
          markImageByBitmap(bg, mark, dest, opts, promise);
        }
      }
    } catch (Exception e) {
      Log.d(IMAGE_MARKER_TAG, "error：" + e.getMessage());
      e.printStackTrace();
      promise.reject("error", e.getMessage(), e);
    }
  }

  private void markImageByBitmap(
    Bitmap bg,
    Bitmap marker,
    String dest,
    MarkImageOptions opts,
    final Promise promise
  ) {
    BufferedOutputStream bos = null;
    Bitmap icon = null;
    try {

      // 原图生成 - start
      int height = bg.getHeight();
      int width = bg.getWidth();

      icon = Utils.getBlankBitmap(width, height);
      //过滤一些

//            if (percent > 1) {
//                prePhoto = Bitmap.createScaledBitmap(prePhoto, width, height, true);
//            }
      Canvas canvas = new Canvas(icon);

      canvas.save();
      canvas.rotate(opts.backgroundImage.rotate);
      canvas.drawBitmap(bg, 0, 0, opts.backgroundImage.applyStyle());
      canvas.restore();
      // 原图生成 - end

      canvas.save();
      canvas.rotate(opts.watermarkImage.rotate);
      if (opts.positionEnum != null) {
        Position pos = Position.getImageRectFromPosition(opts.positionEnum, marker.getWidth(), marker.getHeight(), width, height);
        canvas.drawBitmap(marker, pos.getX(), pos.getY(), opts.watermarkImage.applyStyle());
      } else {
        canvas.drawBitmap(marker, opts.X, opts.Y, opts.watermarkImage.applyStyle());
      }
      canvas.restore();

      if (bg != null && !bg.isRecycled()) {
        bg.recycle();
        System.gc();
      }

      if (marker != null && !marker.isRecycled()) {
        marker.recycle();
        System.gc();
      }

      // 保存
      // canvas.save(Canvas.ALL_SAVE_FLAG);
      canvas.save();
      // 存储
      canvas.restore();
      // export base64
      if (dest.equals(BASE64)) {
        ByteArrayOutputStream base64Stream = new ByteArrayOutputStream();
        icon.compress(Bitmap.CompressFormat.PNG, opts.quality, base64Stream);
        base64Stream.flush();
        base64Stream.close();
        byte[] bitmapBytes = base64Stream.toByteArray();
        String result = Base64.encodeToString(bitmapBytes, Base64.DEFAULT);
        promise.resolve("data:image/png;base64,".concat(result));
      } else {
        bos = new BufferedOutputStream(new FileOutputStream(dest));
        icon.compress(getSaveFormat(opts.saveFormat), opts.quality, bos);
        bos.flush();
        bos.close();
        //保存成功的
        promise.resolve(dest);
      }

    } catch (Exception e) {
      e.printStackTrace();
      promise.reject("error", e.getMessage(), e);
    } finally {
      if (bos != null) {
        try {
          bos.close();
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
      if (icon != null && !icon.isRecycled()) {
        icon.recycle();
        icon = null;
        System.gc();
      }
    }
  }

  /**
   * @param bg
   * @param dest
   * @param opts
   * @param promise
   */
  private void markImageByText(
    Bitmap bg,
    String dest,
    MarkTextOptions opts,
    final Promise promise
  ) {
    BufferedOutputStream bos = null;
    boolean isFinished;
    Bitmap icon = null;
    try {

      int height = bg.getHeight();
      int width = bg.getWidth();

      icon = Utils.getBlankBitmap(width, height);
      //初始化画布 绘制的图像到icon上
      Canvas canvas = new Canvas(icon);
      //建立画笔
      Paint photoPaint = new Paint();
      //获取跟清晰的图像采样
      photoPaint.setDither(true);

      canvas.drawBitmap(bg, 0, 0, photoPaint);

      if (bg != null && !bg.isRecycled()) {
        bg.recycle();
        System.gc();
      }

      Paint textPaint;
      for (TextOptions text : opts.watermarkTexts) {
        //建立画笔
        textPaint = text.applyStyle(this.getReactApplicationContext(), canvas, width, height);
        textPaint.reset();
      }

      if (dest.equals(BASE64)) {
        ByteArrayOutputStream base64Stream = new ByteArrayOutputStream();
        icon.compress(Bitmap.CompressFormat.PNG, opts.quality, base64Stream);
        base64Stream.flush();
        base64Stream.close();
        byte[] bitmapBytes = base64Stream.toByteArray();
        String result = Base64.encodeToString(bitmapBytes, Base64.DEFAULT);
        promise.resolve("data:image/png;base64,".concat(result));
      } else {
        bos = new BufferedOutputStream(new FileOutputStream(dest));
        icon.compress(getSaveFormat(opts.saveFormat), opts.quality, bos);
        bos.flush();
        bos.close();
        //保存成功的
        promise.resolve(dest);
      }


    } catch (Exception e) {
      e.printStackTrace();
      promise.reject("error", e.getMessage(), e);
    } finally {
      isFinished = true;
      if (bos != null) {
        try {
          bos.close();
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
      if (icon != null && !icon.isRecycled()) {
        icon.recycle();
        System.gc();
      }
    }
  }

  /**
   * @param opts
   * @param promise
   */
  @ReactMethod
  public void markWithText(
    ReadableMap opts,
    final Promise promise
  ) {
    MarkTextOptions markOpts = MarkTextOptions.checkParams(opts, promise);
    if (null == markOpts) return;
    try {

      final String dest = generateCacheFilePathForMarker(markOpts.backgroundImage.uri, markOpts.filename, markOpts.saveFormat);

      Log.d(IMAGE_MARKER_TAG, "uri: " + markOpts.backgroundImage.uri);
      Log.d(IMAGE_MARKER_TAG, "src: " + markOpts.backgroundImage.src.toString());

      if (isFrescoImg(markOpts.backgroundImage.uri)) {
        ImageRequest imageRequest = ImageRequest.fromUri(markOpts.backgroundImage.uri);
        if (markOpts.maxSize > 0) {
          setMaxBitmapSize(markOpts.maxSize);
        }
        DataSource<CloseableReference<CloseableImage>> dataSource = getImagePipeline().fetchDecodedImage(imageRequest, null);
        Executor executor = Executors.newSingleThreadExecutor();
        dataSource.subscribe(new BaseBitmapDataSubscriber() {
          @Override
          public void onNewResultImpl(Bitmap bitmap) {
            if (bitmap != null) {
              Bitmap bg = Utils.scaleBitmap(bitmap, markOpts.backgroundImage.scale);

              markImageByText(bg, dest, markOpts, promise);

            } else {
              promise.reject("marker error", "Can't retrieve the file from the src: " + markOpts.backgroundImage.uri);
            }
          }

          @Override
          public void onFailureImpl(DataSource dataSource) {
            promise.reject("error", "Can't request the image from the uri: " + markOpts.backgroundImage.uri, dataSource.getFailureCause());
          }
        }, executor);
      } else {
        int resId = getDrawableResourceByName(markOpts.backgroundImage.uri);
        if (resId == 0) {
          Log.d(IMAGE_MARKER_TAG, "cannot find res");
          promise.reject("error", "Can't get resource by the path: " + markOpts.backgroundImage.uri);
        } else {
          Log.d(IMAGE_MARKER_TAG, "res：" + resId);

          Resources r = this.getResources();
//                    InputStream is = r.openRawResource(resId);
          Bitmap bitmap = BitmapFactory.decodeResource(r, resId);
//                    Bitmap bitmap = BitmapFactory.decodeStream(is);
          Log.d(IMAGE_MARKER_TAG, bitmap.getHeight() + "");
          Bitmap bg = Utils.scaleBitmap(bitmap, markOpts.backgroundImage.scale);
          Log.d(IMAGE_MARKER_TAG, bg.getHeight() + "");

          if (bitmap != null && !bitmap.isRecycled() && markOpts.backgroundImage.scale != 1) {
            bitmap.recycle();
            System.gc();
          }
          markImageByText(bg, dest, markOpts, promise);
        }
      }
    } catch (Exception e) {
      Log.d(IMAGE_MARKER_TAG, "error：" + e.getMessage());
      e.printStackTrace();
      promise.reject("error", e.getMessage(), e);
    }
  }

  @ReactMethod
  public void markWithImage(
    ReadableMap opts,
    final Promise promise) {
    try {
      MarkImageOptions markOpts = MarkImageOptions.checkParams(opts, promise);
      if (null == markOpts) return;

      final String uri = markOpts.backgroundImage.uri;

      final String dest = generateCacheFilePathForMarker(uri, markOpts.filename, markOpts.saveFormat);

      Log.d(IMAGE_MARKER_TAG, uri);
      Log.d(IMAGE_MARKER_TAG, "src: " + markOpts.backgroundImage.src.toString());
      Log.d(IMAGE_MARKER_TAG, "markerSrc: " + markOpts.watermarkImage.src.toString());

      if (isFrescoImg(uri)) {
        ImageRequest imageRequest = ImageRequest.fromUri(uri);
        if (null != markOpts.maxSize && markOpts.maxSize > 0) {
          setMaxBitmapSize(markOpts.maxSize);
        }
        DataSource<CloseableReference<CloseableImage>> dataSource = getImagePipeline().fetchDecodedImage(imageRequest, null);
        Executor executor = Executors.newSingleThreadExecutor();
        dataSource.subscribe(new BaseBitmapDataSubscriber() {
          @Override
          public void onNewResultImpl(Bitmap bitmap) {
            if (bitmap != null) {
              Bitmap bg = Utils.scaleBitmap(bitmap, markOpts.backgroundImage.scale);
              markImage(bg, dest, markOpts, promise);
            } else {
              promise.reject("marker error", "Can't retrieve the file from the src: " + uri);
            }
          }

          @Override
          public void onFailureImpl(DataSource dataSource) {
            promise.reject("error", "Can't request the image from the uri: " + uri, dataSource.getFailureCause());
          }
        }, executor);
      } else {
        int resId = getDrawableResourceByName(uri);
        if (resId == 0) {
          Log.d(IMAGE_MARKER_TAG, "cannot find res");
          promise.reject("error", "Can't get resource by the path: " + uri);
        } else {
          Log.d(IMAGE_MARKER_TAG, "res：" + resId);

          Resources r = this.getResources();
//                    InputStream is = r.openRawResource(resId);
          Bitmap bitmap = BitmapFactory.decodeResource(r, resId);
//                    Bitmap bitmap = BitmapFactory.decodeStream(is);
          Log.d(IMAGE_MARKER_TAG, bitmap.getHeight() + "");
          Bitmap bg = Utils.scaleBitmap(bitmap, markOpts.backgroundImage.scale);
          Log.d(IMAGE_MARKER_TAG, bg.getHeight() + "");

          if (bitmap != null && !bitmap.isRecycled() && markOpts.backgroundImage.scale != 1) {
            bitmap.recycle();
            System.gc();
          }
          markImage(bg, dest, markOpts, promise);
        }
      }
    } catch (Exception e) {
      Log.d(IMAGE_MARKER_TAG, "error：" + e.getMessage());
      e.printStackTrace();
      promise.reject("error", e.getMessage(), e);
    }

  }

  private String generateCacheFilePathForMarker(String imgSavePath, String filename, SaveFormat saveFormat) {
    String cacheDir = this.getReactApplicationContext().getCacheDir().getAbsolutePath();

    if (saveFormat != null && saveFormat == SaveFormat.BASE64) {
      return BASE64;
    }
    String ext = saveFormat != null && (saveFormat == SaveFormat.PNG || saveFormat.equals("png")) ? ".png" : ".jpg";
    if (null != filename) {
      if (filename.endsWith(".jpg") || filename.endsWith(".png"))
        return cacheDir + "/" + filename;
      else
        return cacheDir + "/" + filename + ext;
    } else {
      String name = UUID.randomUUID().toString() + "imagemarker";

      return cacheDir + "/" + name + ext;
    }
  }
}

File: android/src/main/java/com/jimmydaddy/imagemarker/ImageMarkerManager.kt
package com.jimmydaddy.imagemarker

import android.graphics.Bitmap
import android.graphics.Bitmap.CompressFormat
import android.os.Build
import android.util.Base64
import android.util.Log
import androidx.annotation.RequiresApi
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.jimmydaddy.imagemarker.base.Constants.BASE64
import com.jimmydaddy.imagemarker.base.Constants.IMAGE_MARKER_TAG
import com.jimmydaddy.imagemarker.base.ErrorCode
import com.jimmydaddy.imagemarker.base.InvisibleWatermarkOptions
import com.jimmydaddy.imagemarker.base.MarkImageOptions
import com.jimmydaddy.imagemarker.base.MarkTextOptions
import com.jimmydaddy.imagemarker.base.MarkWatermarkOptions
import com.jimmydaddy.imagemarker.base.MarkerError
import com.jimmydaddy.imagemarker.base.Options
import com.jimmydaddy.imagemarker.base.OutputFileName
import com.jimmydaddy.imagemarker.base.SaveFormat
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

/**
 * Created by jimmydaddy on 2017/3/6.
 */
class ImageMarkerManager(private val context: ReactApplicationContext) : NativeImageMarkerSpec(
  context
) {
  private val moduleScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private val markerJobLimiter = MarkerJobLimiter(parallelism = 1)
  private val markerJobs = ConcurrentHashMap<String, kotlinx.coroutines.Job>()

  override fun invalidate() {
    markerJobs.values.forEach { it.cancel() }
    markerJobs.clear()
    moduleScope.cancel()
    super.invalidate()
  }

  private fun getSaveFormat(saveFormat: SaveFormat?): CompressFormat {
    return when (saveFormat) {
      SaveFormat.PNG -> CompressFormat.PNG
      SaveFormat.WEBP ->
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
          CompressFormat.WEBP_LOSSY
        } else {
          @Suppress("DEPRECATION")
          CompressFormat.WEBP
        }
      else -> CompressFormat.JPEG
    }
  }

  private fun launchMarkerJob(
    jobId: String,
    promise: Promise,
    block: suspend () -> String
  ) {
    val job = moduleScope.launch(start = CoroutineStart.LAZY) {
      try {
        promise.resolve(markerJobLimiter.run(block))
      } catch (e: CancellationException) {
        Log.d(IMAGE_MARKER_TAG, "marker job cancelled")
        promise.reject("ABORTED", "Image marker operation was aborted", e)
      } catch (error: OutOfMemoryError) {
        val markerError = MarkerError(
          ErrorCode.RENDER_FAILED,
          "Unable to complete marker image"
        ).apply { initCause(error) }
        Log.d(IMAGE_MARKER_TAG, "error: " + markerError.message)
        promise.rejectMarkerError(markerError)
      } catch (e: Exception) {
        Log.d(IMAGE_MARKER_TAG, "error: " + e.message)
        promise.rejectMarkerError(e)
      } finally {
        markerJobs.remove(jobId)
      }
    }
    markerJobs[jobId] = job
    job.start()
  }

  private fun Promise.rejectMarkerError(error: Exception) {
    error.printStackTrace()
    if (error is MarkerError) {
      reject(error.getErrorCode(), error.getErrMsg(), error)
    } else {
      reject("error", error.message, error)
    }
  }

  private suspend fun markImageByBitmap(
    bg: Bitmap,
    markers: List<Bitmap>,
    dest: String,
    opts: MarkImageOptions
  ): String {
    return withContext(Dispatchers.Default) {
      OwnedResourcePipeline.run(
        inputs = listOf(bg).plus(markers),
        releaseInput = ::recycleBitmap,
        render = {
          ImageMarkerRenderer.renderImageWatermarks(
            bg,
            markers,
            opts,
            recycleMarkerBitmaps = false
          )
        },
        // The renderer owns a distinct output bitmap. Inputs are released before encoding,
        // which is often the second-largest memory peak in the pipeline.
        encode = { icon -> writeResult(icon, dest, opts) },
        releaseOutput = ::recycleBitmap
      )
    }
  }

  private suspend fun markImageByText(
    bg: Bitmap,
    dest: String,
    opts: MarkTextOptions
  ): String {
    return withContext(Dispatchers.Default) {
      OwnedResourcePipeline.run(
        inputs = listOf(bg),
        releaseInput = ::recycleBitmap,
        render = {
          ImageMarkerRenderer.renderTextWatermarks(bg, opts, reactApplicationContext)
        },
        encode = { icon -> writeResult(icon, dest, opts) },
        releaseOutput = ::recycleBitmap
      )
    }
  }

  private suspend fun markImageByWatermarks(
    bg: Bitmap,
    markers: List<Bitmap>,
    dest: String,
    opts: MarkWatermarkOptions
  ): String {
    return withContext(Dispatchers.Default) {
      OwnedResourcePipeline.run(
        inputs = listOf(bg).plus(markers),
        releaseInput = ::recycleBitmap,
        render = {
          ImageMarkerRenderer.renderWatermarks(
            bg,
            markers,
            opts,
            context,
            recycleMarkerBitmaps = false
          )
        },
        encode = { icon -> writeResult(icon, dest, opts) },
        releaseOutput = ::recycleBitmap
      )
    }
  }

  private suspend fun embedInvisibleWatermark(
    bg: Bitmap,
    dest: String,
    opts: InvisibleWatermarkOptions
  ): String {
    return withContext(Dispatchers.Default) {
      OwnedResourcePipeline.run(
        inputs = listOf(bg),
        releaseInput = ::recycleBitmap,
        render = {
          InvisibleWatermark.embed(
            bg,
            opts.requirePayload(),
            opts.key,
            opts.strength
          )
        },
        encode = { output -> writeResult(output, dest, opts) },
        releaseOutput = ::recycleBitmap
      )
    }
  }

  private suspend fun writeResult(
    icon: Bitmap,
    dest: String,
    opts: Options
  ): String = withContext(Dispatchers.IO) {
    try {
      if (dest == BASE64) {
        return@withContext encodeBase64(icon, opts)
      }

      AtomicFileWriter.write(File(dest)) { stream ->
        if (!icon.compress(getSaveFormat(opts.saveFormat), opts.quality, stream)) {
          throw IOException("Failed to encode marker image")
        }
      }
      dest
    } catch (error: OutOfMemoryError) {
      throw encodingOutOfMemory(error)
    }
  }

  private fun encodeBase64(
    icon: Bitmap,
    opts: Options
  ): String {
    try {
      val bitmapBytes = ByteArrayOutputStream().use { stream ->
        if (!icon.compress(CompressFormat.PNG, opts.quality, stream)) {
          throw IOException("Failed to encode marker image")
        }
        stream.flush()
        stream.toByteArray()
      }
      val result = Base64.encodeToString(bitmapBytes, Base64.DEFAULT)
      return "data:image/png;base64,$result"
    } catch (error: OutOfMemoryError) {
      throw encodingOutOfMemory(error)
    }
  }

  private fun encodingOutOfMemory(error: OutOfMemoryError): MarkerError {
    return MarkerError(
      ErrorCode.RENDER_FAILED,
      "Unable to encode marker image"
    ).apply { initCause(error) }
  }

  private fun recycleBitmaps(bitmaps: List<Bitmap>) {
    for (bitmap in bitmaps) {
      recycleBitmap(bitmap)
    }
  }

  private fun recycleBitmap(bitmap: Bitmap?) {
    if (bitmap != null && !bitmap.isRecycled) {
      bitmap.recycle()
    }
  }

  /**
   * @param opts
   * @param promise
   */
  @RequiresApi(Build.VERSION_CODES.N)
  @ReactMethod
  override fun embedInvisible(
    options: ReadableMap,
    promise: Promise
  ) {
    val opts = InvisibleWatermarkOptions.checkEmbed(options, promise) ?: return
    launchMarkerJob(opts.jobId, promise) {
      val bitmaps = MarkerImageLoader(context, opts.maxSize).loadImages(
        listOf(opts.backgroundImage)
      )
      try {
        val dest = generateCacheFilePathForMarker(opts.filename, opts.saveFormat)
        embedInvisibleWatermark(bitmaps[0], dest, opts)
      } finally {
        recycleBitmaps(bitmaps)
      }
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  @ReactMethod
  override fun detectInvisible(
    options: ReadableMap,
    promise: Promise
  ) {
    val opts = InvisibleWatermarkOptions.checkDetect(options, promise) ?: return
    launchMarkerJob(opts.jobId, promise) {
      val bitmaps = MarkerImageLoader(context, opts.maxSize).loadImages(
        listOf(opts.backgroundImage)
      )
      try {
        withContext(Dispatchers.Default) {
          InvisibleWatermark.detect(
            bitmaps[0],
            opts.key,
            opts.strength,
            opts.search
          ).toJson()
        }
      } finally {
        recycleBitmaps(bitmaps)
      }
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  @ReactMethod
  override fun markWithText(
    options: ReadableMap,
    promise: Promise
  ) {
    val markOpts = MarkTextOptions.checkParams(options, promise) ?: return
    Log.d(IMAGE_MARKER_TAG, "uri: " + markOpts.backgroundImage.uri)
    Log.d(IMAGE_MARKER_TAG, "src: " + markOpts.backgroundImage.src.toString())
    launchMarkerJob(markOpts.jobId, promise) {
      val bitmaps = MarkerImageLoader(context, markOpts.maxSize).loadImages(
        listOf(
          markOpts.backgroundImage,
        )
      )
      try {
        val bg = bitmaps[0]
        val dest = generateCacheFilePathForMarker(markOpts.filename, markOpts.saveFormat)
        markImageByText(bg, dest, markOpts)
      } finally {
        recycleBitmaps(bitmaps)
      }
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  @ReactMethod
  override fun markWithImage(
    options: ReadableMap,
    promise: Promise
  ) {
    val markOpts = MarkImageOptions.checkParams(options, promise) ?: return
    launchMarkerJob(markOpts.jobId, promise) {
      val markers = markOpts.watermarkImages.map { it.imageOption }
      val concatenatedArray = listOf(
        markOpts.backgroundImage,
      ).plus(markers)
      val bitmaps = MarkerImageLoader(context, markOpts.maxSize).loadImages(
        concatenatedArray,
        listOf(true).plus(List(markers.size) { false })
      )
      try {
        val bg = bitmaps[0]
        val markerBitmaps = bitmaps.drop(1)
        val dest = generateCacheFilePathForMarker(markOpts.filename, markOpts.saveFormat)
        markImageByBitmap(bg, markerBitmaps, dest, markOpts)
      } finally {
        recycleBitmaps(bitmaps)
      }
    }
  }

  @RequiresApi(Build.VERSION_CODES.N)
  @ReactMethod
  override fun markWithWatermarks(
    options: ReadableMap,
    promise: Promise
  ) {
    val markOpts = MarkWatermarkOptions.checkParams(options, promise) ?: return
    launchMarkerJob(markOpts.jobId, promise) {
      val markers = markOpts.imageLayers.map { it.imageOptions.imageOption }
      val concatenatedArray = listOf(
        markOpts.backgroundImage,
      ).plus(markers)
      val bitmaps = MarkerImageLoader(context, markOpts.maxSize).loadImages(
        concatenatedArray,
        listOf(true).plus(List(markers.size) { false })
      )
      try {
        val bg = bitmaps[0]
        val markerBitmaps = bitmaps.drop(1)
        val dest = generateCacheFilePathForMarker(markOpts.filename, markOpts.saveFormat)
        markImageByWatermarks(bg, markerBitmaps, dest, markOpts)
      } finally {
        recycleBitmaps(bitmaps)
      }
    }
  }

  @ReactMethod
  override fun cancel(
    jobId: String,
    promise: Promise
  ) {
    val job = markerJobs.remove(jobId)
    if (job == null) {
      promise.resolve(false)
      return
    }
    job.cancel(CancellationException("Image marker job $jobId was cancelled"))
    promise.resolve(true)
  }

  private fun generateCacheFilePathForMarker(
    filename: String?,
    saveFormat: SaveFormat?
  ): String {
    val cacheDir = this.reactApplicationContext.cacheDir.absolutePath
    if (saveFormat != null && saveFormat === SaveFormat.BASE64) {
      return BASE64
    }
    val generatedName = UUID.randomUUID().toString() + "_image_marker"
    val outputFileName = OutputFileName.resolve(filename, saveFormat, generatedName)
    return "$cacheDir/$outputFileName"
  }

  companion object {
    const val NAME = "ImageMarker"
  }
}
